src/high-level/constructors.lisp

```diff
@@ -0,0 +1,173 @@
+;;;; constructors.lisp
+;;;;
+;;;; Author: Cole Scott
+
+(in-package #:magicl)
+
+(defconstant +default-tensor-type+ 'double-float)
+
+(defgeneric make-tensor (class shape &key initial-element layout storage)
+  (:documentation "Make a dense tensor with elements of the specified type"))
+
+(defun empty (shape &key (type +default-tensor-type+) layout)
+  "Create a tensor without intializing the contents of the storage
+
+If TYPE is not specified then +DEFAULT-TENSOR-TYPE+ is used.
+LAYOUT specifies the internal storage represenation ordering of the returned tensor.
+The tensor specialized on the specified SHAPE and TYPE."
+  (policy-cond:with-expectations (> speed safety)
+      ((type shape shape))
+    (let ((tensor-type (infer-tensor-type type shape nil)))
+      (make-tensor tensor-type shape :layout layout))))
+
+(defun const (const shape &key type layout)
+  "Create a tensor with the specified SHAPE with each element being set to CONST
+
+If TYPE is not specified then it is inferred from the type of CONST.
+LAYOUT specifies the internal storage represenation ordering of the returned tensor.
+The tensor is specialized on SHAPE and TYPE."
+  (policy-cond:with-expectations (> speed safety)
+      ((type shape shape))
+    (let ((tensor-class (infer-tensor-type type shape const)))
+      (make-tensor tensor-class shape :layout layout :initial-element const))))
+
+(defun rand (shape &key (type +default-tensor-type+) layout distribution)
+  "Create tensor with random elements from DISTRIBUTION
+
+DISTRIBUTION is a function with no arguments which returns a value for the element.
+If DISTRIBUTION is not specified then CL:RANDOM is used. In the case that TYPE is complex, CL:RANDOM is called for each component.
```

Comment:
Change the documentation string and remove the `cl:random` stuff and say instead that by default a random number in the unit interval will be drawn (similarly for the real and imaginary parts of a complex number). After all, `cl:random` cannot be called without arguments anyway.